integrations/operator/controllers/resources/teleport_reconciler.go

```diff
@@ -176,22 +203,46 @@ func (r TeleportResourceReconciler[T, K]) Delete(ctx context.Context, obj kclien
 
 // Reconcile allows the TeleportResourceReconciler to implement the reconcile.Reconciler interface
 func (r TeleportResourceReconciler[T, K]) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	kubeResource := newKubeResource[T, K]()
-	return r.Do(ctx, req, kubeResource)
+	obj, err := GetUnstructuredObjectFromGVK(r.gvk)
+	if err != nil {
+		return ctrl.Result{}, trace.Wrap(err, "creating object in which the CR will be unmarshalled")
+	}
+	return r.Do(ctx, req, obj)
 }
 
 // SetupWithManager have a controllerruntime.Manager run the TeleportResourceReconciler
 func (r TeleportResourceReconciler[T, K]) SetupWithManager(mgr ctrl.Manager) error {
-	kubeResource := newKubeResource[T, K]()
+	// The TeleportResourceReconciler uses unstructured objects because of a silly json marshalling
+	// issue. Teleport's utils.String is a list of strings, but marshalls as a single string if there's a single item.
+	// This is a questionable design as it breaks the openapi schema, but we're stuck with it. We had to relax openapi
+	// validation in those CRD fields, and use an unstructured object for the client, else JSON unmarshalling fails.
+	obj, err := GetUnstructuredObjectFromGVK(r.gvk)
+	if err != nil {
+		return trace.Wrap(err, "creating the model object for the manager watcher/client")
+	}
 	return ctrl.
 		NewControllerManagedBy(mgr).
-		For(kubeResource).
+		For(obj).
 		WithEventFilter(
 			buildPredicate(),
 		).
 		Complete(r)
 }
 
+// gvkFromScheme looks up the GVK from the a runtime scheme.
+// The structured type must have been registered before in the scheme. This function is used when you have a structured
+// type, a scheme containing this structured type, and want to build an unstructured object for the same GVK.
+// The TeleportResourceReconciler unpacks in unstructured types first, thenm
+func gvkFromScheme[T TeleportResource, K TeleportKubernetesResource[T]](scheme *runtime.Scheme) (schema.GroupVersionKind, error) {
+	structuredObj := newKubeResource[T, K]()
+	gvks, _, err := scheme.ObjectKinds(structuredObj)
+	if err != nil {
+		return schema.GroupVersionKind{}, trace.Wrap(err, "looking up gvk in scheme for type %T", structuredObj)
+	}
+	// Change this when we'll add multi version support
```

Comment:
This will always map 1-> 1 even in multi version support. 
We should also panic/return an error if len is not 1 since it's required to be